### PR TITLE
tstest/integration/vms: test vms as they are ready

### PR DESCRIPTION
**Pull request by Xe**
Instead of testing all the VMs at once when they are all ready, this
patch changes the testing logic so that the vms are tested as soon as
they register with testcontrol. Also limit the amount of VM ram used at
once with the `-ram-limit` flag. That uses a semaphore to guard resource
use.

Updates #1988

Signed-off-by: Christine Dodrill <xe@tailscale.com>

### CODE REVIEW

**Comment by Xe** [comment]
```
--- PASS: TestVMIntegrationEndToEnd (137.53s)
    --- PASS: TestVMIntegrationEndToEnd/do (0.00s)
        --- SKIP: TestVMIntegrationEndToEnd/do/opensuse-leap-15-1 (0.00s)
        --- PASS: TestVMIntegrationEndToEnd/do/debian-10 (22.64s)
            --- PASS: TestVMIntegrationEndToEnd/do/debian-10/wait-for-start (17.00s)
        --- PASS: TestVMIntegrationEndToEnd/do/ubuntu-16-04 (24.38s)
            --- PASS: TestVMIntegrationEndToEnd/do/ubuntu-16-04/wait-for-start (19.00s)
        --- PASS: TestVMIntegrationEndToEnd/do/ubuntu-21-04 (31.08s)
            --- PASS: TestVMIntegrationEndToEnd/do/ubuntu-21-04/wait-for-start (27.00s)
        --- PASS: TestVMIntegrationEndToEnd/do/ubuntu-20-10 (31.78s)
            --- PASS: TestVMIntegrationEndToEnd/do/ubuntu-20-10/wait-for-start (28.00s)
        --- PASS: TestVMIntegrationEndToEnd/do/opensuse-tumbleweed (32.74s)
            --- PASS: TestVMIntegrationEndToEnd/do/opensuse-tumbleweed/wait-for-start (30.01s)
        --- PASS: TestVMIntegrationEndToEnd/do/ubuntu-18-04 (34.86s)
            --- PASS: TestVMIntegrationEndToEnd/do/ubuntu-18-04/wait-for-start (31.00s)
        --- PASS: TestVMIntegrationEndToEnd/do/debian-9 (50.60s)
            --- PASS: TestVMIntegrationEndToEnd/do/debian-9/wait-for-start (17.00s)
        --- PASS: TestVMIntegrationEndToEnd/do/ubuntu-20-04 (52.14s)
            --- PASS: TestVMIntegrationEndToEnd/do/ubuntu-20-04/wait-for-start (26.00s)
        --- PASS: TestVMIntegrationEndToEnd/do/amazon-linux (64.47s)
            --- PASS: TestVMIntegrationEndToEnd/do/amazon-linux/wait-for-start (22.00s)
        --- PASS: TestVMIntegrationEndToEnd/do/centos-7 (67.50s)
            --- PASS: TestVMIntegrationEndToEnd/do/centos-7/wait-for-start (14.00s)
        --- PASS: TestVMIntegrationEndToEnd/do/centos-8 (86.79s)
            --- PASS: TestVMIntegrationEndToEnd/do/centos-8/wait-for-start (19.00s)
        --- PASS: TestVMIntegrationEndToEnd/do/fedora-34 (74.66s)
            --- PASS: TestVMIntegrationEndToEnd/do/fedora-34/wait-for-start (60.00s)
        --- PASS: TestVMIntegrationEndToEnd/do/opensuse-leap-15-2 (137.31s)
            --- PASS: TestVMIntegrationEndToEnd/do/opensuse-leap-15-2/wait-for-start (103.00s)
```

File: tstest/integration/vms/vms_test.go [hunk 1]
```diff
@@ -280,6 +282,7 @@ func mkVM(t *testing.T, n int, d Distro, sshKey, hostURL, tdir string) func() {
 		"-drive", driveArg,
 		"-cdrom", filepath.Join(tdir, d.name, "seed", "seed.iso"),
 		"-vnc", fmt.Sprintf(":%d", n),
+		"-smbios", "type=1,serial=ds=nocloud;h=" + d.name,
```
Comment (danderson) [inline tstest/integration/vms/vms_test.go:285]: Why this?

File: tstest/integration/vms/vms_test.go [hunk 2]
```diff
@@ -424,140 +428,126 @@ func TestVMIntegrationEndToEnd(t *testing.T) {
...
-						break
-					}
-
-					time.Sleep(5 * time.Second)
-				}
-
-				if !working {
-					t.Fatalf("can't connect to %s, tried %d times", hostport, maxRetries)
-				}
+		time.Sleep(10 * time.Second)
```
Comment (danderson) [inline tstest/integration/vms/vms_test.go:507]: 10s is maybe a bit slow here. Make it 500ms and use a deadline instead of a fixed retry count?

**Review by danderson — APPROVED** [review]

File: tstest/integration/vms/vms_test.go (hunk 1 above)
Comment (Xe) [reply tstest/integration/vms/vms_test.go:285]: The [NoCloud](https://cloudinit.readthedocs.io/en/latest/topics/datasources/nocloud.html) datastore really prefers having a SMBIOS variable telling it where to look for data. In this case we are just telling it to read from the cloud-init seed, however when I implement the network metadata storage part of #1988 it will be how we tell where cloud-init should read metadata from. This was also an ill-fated attempt to get OpenSUSE 15.1 working (my VM puppeting setup that uses the same basic stack as this testsuite is able to work with it, and it sets that smbios variable, however it's just not working and IDK why; I'll figure that out later).

tl;dr: It's all about correctness in how I do satanic things with cloud-init.

File: tstest/integration/vms/vms_test.go (hunk 2 above)
Comment (Xe) [reply tstest/integration/vms/vms_test.go:510]: This is intended to work around CentOS brokenness :( CentOS will slay all SSH sessions on transition to multi-user.target. I can make it into a deadline loop in a later patch.
